src/workerd/api/web-socket.c++

```diff
@@ -794,47 +785,70 @@ kj::Promise<void> WebSocket::pump(
-        -> kj::Promise<void> {
-      jsg::Lock& lock = wLock;
-      auto isolate = lock.v8Isolate;
-      auto& native = *farNative;
-      KJ_SWITCH_ONEOF(message) {
-        KJ_CASE_ONEOF(text, kj::String) {
-          dispatchEventImpl(lock, jsg::alloc<MessageEvent>(isolate, lock.wrapString(text)));
-        }
-        KJ_CASE_ONEOF(data, kj::Array<byte>) {
-          dispatchEventImpl(lock,
-              jsg::alloc<MessageEvent>(isolate, lock.wrapBytes(kj::mv(data))));
-        }
-        KJ_CASE_ONEOF(close, kj::WebSocket::Close) {
-          native.closedIncoming = true;
-          dispatchEventImpl(lock, jsg::alloc<CloseEvent>(close.code, kj::mv(close.reason), true));
-          // Native WebSocket no longer needed; release.
-          tryReleaseNative(lock);
-          return kj::READY_NOW;
-        }
+      context.getLimitEnforcer().topUpActor();
+      KJ_IF_MAYBE(a, context.getActor()) {
+        auto size = countBytesFromMessage(message);
+        a->getMetrics().receivedWebSocketMessage(size);
       }
 
-      return readLoop(ws);
-    });
-  });
+      // Re-enter the context with context.run(). This is arguably a bit unusual compared to other
+      // I/O which is delivered by return from context.awaitIo(), but the difference here is that we
+      // have a long stream of events over time. It makes sense to use context.run() each time a new
+      // event arrives.
+      // TODO(cleanup): The way context.run is defined, a capturing lambda is required here, which
+      // is a bit unfortunate. We could simply things somewhat with a variation that would allow
+      // something like context.run(handleMessage, *this, kj::mv(message)) where the acquired lock,
+      // and the additional arguments are passed into handleMessage, avoiding the need for the
+      // lambda here entirely.
+      auto result = co_await context.run([this, message=kj::mv(message)](auto& wLock) mutable {
+        auto& native = *farNative;
+        jsg::Lock& js = wLock;
+        KJ_SWITCH_ONEOF(message) {
+          KJ_CASE_ONEOF(text, kj::String) {
+            dispatchEventImpl(js, jsg::alloc<MessageEvent>(js.v8Isolate, js.wrapString(text)));
+          }
+          KJ_CASE_ONEOF(data, kj::Array<byte>) {
+            dispatchEventImpl(js, jsg::alloc<MessageEvent>(js.v8Isolate,
+                                                           js.wrapBytes(kj::mv(data))));
+          }
+          KJ_CASE_ONEOF(close, kj::WebSocket::Close) {
+            native.closedIncoming = true;
+            dispatchEventImpl(js, jsg::alloc<CloseEvent>(close.code, kj::mv(close.reason), true));
+            // Native WebSocket no longer needed; release.
+            tryReleaseNative(js);
+            return false;
+          }
+        }
+
+        return true;
+      });
+
+      if (!result) co_return nullptr;
+    }
+    KJ_UNREACHABLE;
+  } catch (...) {
+    co_return kj::getCaughtExceptionAsKj();
+  }
```

Comment:
Isn't this the same as having no `try` at all?

Reply:
I'm not too familiar with `getCaughtExceptionAsKj()`, but this pattern was recommended by @kentonv... as I understand it, this should catch any exception thrown and coerce it into a kj::Exception. This method is designed to intentionally fulfill the promise with the exception or nothing.